### DISCUSSION TITLE
How to implement fleet minimisation?

### DISCUSSION THREAD

**Discussion by mirtiza09** [PyVRP/PyVRP#632] [Q&A] [answered]
In the PyVRP paper from 'INFORMS Journal on Computing', the following has been written:

> For both CVRP and VRPTW, PyVRP supports minimising distances, not the number of vehicles used. A simple procedure to support fleet minimisation can be developed on top of PyVRP, by first finding a feasible solution with some number of vehicles, and then removing one vehicle at a time until no feasible solution is found

Are the writers referring to Nagata et al.'s work from 2010 (https://doi.org/10.1016/j.orl.2009.04.006)?

If yes, it is not clear to me how I can implement that in PyVRP from the paper or documentation. Any help in this regard would be appreciated. Also, I would love to contribute to PyVRP for multi-objective optimisation (i.e. Number of routes AND Total distance), but I was wondering if it is even possible or feasible to contribute this.

**Comment by N-Wouda** [comment, marked as answer]
> Are the writers referring to Nagata et al.'s work from 2010 (https://doi.org/10.1016/j.orl.2009.04.006)?

Not on purpose. This idea has been (re)invented many times.

>If yes, it is not clear to me how I can implement that in PyVRP from the paper or documentation. Any help in this regard would be appreciated.

I'd run PyVRP with a given fleet size until you find a first feasible solution, or a time/iteration limit has been reached. If the solution is feasible, reduce the fleet size by one (or to the number of vehicles used in the feasible solution minus one, whichever is smaller) and try again. If the time/iteration limit has been reached, return the fleet size you know results in a feasible solution.

**Reply by N-Wouda** [reply]
You might also appreciate https://github.com/PyVRP/PyVRP/pull/638.

**Comment by mirtiza09** [comment]
@N-Wouda Any timelines for the release of 0.10 version, through which we can implement minimise_fleet? Is there any other way we can download that version ((pip install --upgrade pyvrp) installs 0.9)?

**Reply by N-Wouda** [reply]
Your best bet would be to install from source for now (explained [here](https://pyvrp.org/setup/installation.html#installing-from-source)). I expect we'll release v0.10 late next month, but that depends a bit on my currently fairly limited availability.
